homeassistant/components/subaru/config_flow.py

```diff
@@ -0,0 +1,170 @@
+        except CannotConnect:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "connection_error"},
+                description_placeholders={},
+            )
+        except InvalidAuth:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "invalid_credentials"},
+                description_placeholders={},
+            )
+        return self.async_create_entry(title=user_input[CONF_USERNAME], data=info)
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return OptionsFlowHandler(config_entry)
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle a option flow for Subaru."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry):
+        """Initialize options flow."""
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None):
+        """Handle options flow."""
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        data_schema = vol.Schema(
+            {
+                vol.Optional(
+                    CONF_SCAN_INTERVAL,
+                    default=self.config_entry.options.get(
+                        CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+                    ),
+                ): vol.All(cv.positive_int, vol.Clamp(min=MIN_SCAN_INTERVAL)),
+                vol.Optional(
+                    CONF_HARD_POLL_INTERVAL,
+                    default=self.config_entry.options.get(
+                        CONF_HARD_POLL_INTERVAL, DEFAULT_HARD_POLL_INTERVAL
+                    ),
+                ): vol.All(cv.positive_int, vol.Clamp(min=MIN_HARD_POLL_INTERVAL)),
+            }
+        )
+        return self.async_show_form(step_id="init", data_schema=data_schema)
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    websession = aiohttp_client.async_get_clientsession(hass)
+    now = datetime.now()
+    if not data.get(CONF_DEVICE_ID):
+        data[CONF_DEVICE_ID] = int(now.timestamp())
+    date = now.strftime("%Y-%m-%d")
+    device_name = "Home Assistant: Added " + date
+    try:
+        controller = SubaruAPI(
+            websession,
+            username=data[CONF_USERNAME],
+            password=data[CONF_PASSWORD],
+            device_id=data[CONF_DEVICE_ID],
+            pin=data[CONF_PIN],
+            device_name=device_name,
+        )
+        if await controller.connect(test_login=True):
+            _LOGGER.debug("Successfully authenticated and authorized with Subaru API")
+        else:
+            raise SubaruException("Unknown Error")
```

Comment:
Since this will be caught a couple of lines down and raised as a `CannotConnect`, why not simply raise `CannotConnect` here?

Reply:
Agreed, config flow exceptions were redone to just make use of what is provided by the subarulink module.